### DISCUSSION TITLE
Open Networks

### DISCUSSION THREAD

**Discussion by aluminum-ice** [RaspAP/raspap-webgui#1039] [Q&A] [answered]
I'm using RaspAP with two WiFi networks: wlan0 for the AP and wlan1 as the client connecting to a network with access to the Internet. This works great for any client network that uses a password. But not for open networks like Starbucks, McDonalds or on an airplane. The WiFi client page on RaspAP states it doesn't support connecting to WEP but does that also mean I cannot connect to any open network?

**Comment by billz** [comment]
Good question. Most open wireless networks including, I suspect, the ones you mention use some form of captive portal to manage access. 

When a wireless client device connects to one of these open SSIDs, the captive portal detection (CPD) mechanism will open a splash page to authenticate the client. Once the user on the client device has completed the splash page actions, the page will usually do a redirect with a query string containing a token. For security, the remote server expects the same valid token it allocated when the client received the initial HTTP request. If the token is valid, the portal "authenticates" the client and allows access to the internet.

This is a fairly simple process when the client is a mobile phone or laptop. With RaspAP in the middle it becomes quite a bit more complex. ie., the portal's HTTP response must be captured and handled in one of several ways. The user auth process may have multiple steps (for example, requiring an email address or phone number to receive an access code) and each would need to be brokered by RaspAP. Assuming that all goes well, a persistent token (ie., a "cookie") would need to be managed by RaspAP, else the client would be forced to authenticate all over again.

FWIW, I'm not aware of any commercial travel-type routers (Netgear, TP-Link, etc) that support captive portal logins. 

tl;dr: most open networks use captive portals, which are non-trivial for travel routers to deal with.

**Reply by aluminum-ice** [reply]
Thank you for the detailed response! 

Before using RaspAP I built myself similar functionality that worked with captive portals such as Starbucks by following [these instructions](https://blog.thewalr.us/2017/09/26/raspberry-pi-zero-w-simultaneous-ap-and-managed-mode-wifi/) for a Pi Zero that did simultaneous AP and managed mode WiFi. Basically I would edit wpa_supplicant to add the network details, bring up a connection for the managed mode device (eg, wlan0), and then connect my computer to ap0 as the AP with traffic routed between them. So, what would happen is the Pi Zero would get an IP address on wlan0, and when I used my computer's browser to access a site like cnn.com, the captive portal would be displayed for me to log in as you say. It worked but was awkward. 

With RaspAP it seems the client device wlan1 doesn't get assigned any IP address as the first step unless there is a password on the network it's connecting to. If the network is password protected, the RaspAP gets an IP address and then the captive portal is triggered. Eg, A local non-chain coffee shop provides internet access that uses both a password and a captive portal. You use the password to connect to the network and then when you try to browse cnn.com the captive portal comes up. In that situation RaspAP handles it correctly. But if no password is assigned, it doesn't work. Does my example make sense?

**Reply by FallenChromium** [reply]
Actually, most of the captive portals I've seen are "remembering" the client based on the MAC address, which de-couples the device from which the info was entered and the device which is routing the traffic (i.e RaspAP). The problem I've faced with is that the default config (I believe? Still haven't looked properly into this) uses 1.1.1.1 as a DNS server, so with my (wlan0 as a client + wlan1 as a hotspot + wireguard) setup, the captive portal page never appears with the captive portal checks on either client, and even when I do it directly on the Pi, AND EVEN when I know the address of the login page, I can't reach it because of some upstream DNS problem. If that tricky thing could be circumvented, I think many captive portals would start working properly with RaspAP

**Reply by aluminum-ice** [reply]
> Actually, most of the captive portals I've seen are "remembering" the client based on the MAC address, which de-couples the device from which the info was entered and the device which is routing the traffic (i.e RaspAP). The problem I've faced with is that the default config (I believe? Still haven't looked properly into this) uses 1.1.1.1 as a DNS server, so with my (wlan0 as a client + wlan1 as a hotspot + wireguard) setup, the captive portal page never appears with the captive portal checks on either client, and even when I do it directly on the Pi, AND EVEN when I know the address of the login page, I can't reach it because of some upstream DNS problem. If that tricky thing could be circumvented, I think many captive portals would start working properly with RaspAP

Are you able to get wlan0 to be assigned an IP address (as the client) via RaspAP for an open network? I don't get any IP address assigned unless the network has a password assigned. If it does, using the browser to go to the router's address usually brings up the captive portal for me.

**Reply by FallenChromium** [reply]
I've set up a DHCP client on wlan0 interface using PIXEL widget in the Raspberry's GUI, but my issue here is that 1.1.1.1 is used as the default DNS (and I am not sure how can I check if upstream servers are used), and I'm also using Wireguard always-on, which can surely prevent captive portal from working (but it's the way I access my RPi through cellular networks, so disabling it is not really an option).

**Reply by billz** [reply]
@FallenChromium upstream DNS servers can be managed on RaspAP's DHCP server > Advanced tab.

Re: open networks in general, I will open a PR to address this. Bear in mind captive portal support probably depends in large part on the specific implementation (there are many) and there's a likelihood many will fail to connect. Still with exploring, IMO. 

IIRC, OpenWrt has a wlan connection manager package called Travelmate developed specifically to address this.

**Reply by billz** [reply]
@shinigami-78 this has been implemented in #1050. thanks for the suggestion

**Comment by feh123** [comment]
I never went to step 6 in the following: https://www.makeuseof.com/tag/raspberry-pi-vpn-travel-router/ although I did get the RPI WRT router to work as described on the web site. It's not an easy set up as I recall. 
Luckily I already had RaspAP working very well and much easier then openwrt.

**Comment by aluminum-ice** [comment, marked as answer]
I wanted to share that the latest Insider version v2.8.0 works great now with public wifi spots! I've been able to connect to every public wifi I've tried flawlessly using two wireless connections (onboard and external). Thank you for considering and implementing this change.

**Reply by billz** [reply]
Great! Just as a note for readers, this is implemented in the main RaspAP repo as well (not exclusive to Insiders).
Thanks again for raising this discussion.

**Reply by MrGokilz** [reply]
I use RaspAp v3.0.4 and have a Problem to Connect to Open network.If i try to Connect the Network, I get Always this Error MSG: WPA Passphrase must be between 8 and 63 Charakters.

What should i do, so that i can Connect the WiFi Client to Open Network?

**Reply by crystalstyl** [reply]
> Great! Just as a note for readers, this is implemented in the main RaspAP repo as well (not exclusive to Insiders).
> 
> Thanks again for raising this discussion.

Hi, Has this function already been implemented in the normal version? 
I have the latest version, but I can't log into an "open" network? 
Regards

**Reply by billz** [reply]
@crystalstyl can you be more specific?

**Reply by Moving-Electrons** [reply]
@billz , I’m experiencing the same issue described above. As described in the original post, I have two wireless interfaces: wlan1 for other devices to connect to RaspAP and wlan0 for RaspAp to connect to WiFi networks. RaspAp shows the “open networks” (I.e. those using a captive portal) in the WiFi Client section, but when attempting to connect with no password, I get an error saying that the password needs to be more than 8 characters.

How should I connect to networks not requiring a password that use a Captive Portal to authenticate?.

I’m using the latest version of RaspAp.

Any help will be greatly appreciated. 

Thanks,

**Comment by sakuwa** [comment]
Is there any changes? Im on vacation and tried to connect to an open wlan (nmcli scan shows: security : -- , wpa flags: --,...)

In the gui it's told me this wlan is a wep and can't connected or needs a passphrase. But with my mobile device I can connect to the WiFi. 

I tested at home to connect to a WiFi but only with a wpa2 and it works. 

Any ideas?

**Comment by billz** [comment]
See #1588. Resolved by #1627. Also applied in: https://github.com/RaspAP/raspap-webgui/blob/db9ca93bfbe00308b07819de6819684244a4d488/src/RaspAP/Networking/Hotspot/WiFiManager.php#L424

**Reply by sakuwa** [reply]
Sorry for my double post. Didn't find this thread today 

Mmh so I need to configure this in wpa_supplicant for myself? Otherwise I can't add this open network in webui. Thanks

**Comment by sakuwa** [comment]
Sorry im searching on my mobilephone for a Solution. 

I find something Like this: 

f (strtolower($protocol) === 'open') {
    $commands[] = "sudo wpa_cli -i $iface set_network $netid key_mgmt NONE";
    $commands[] = "sudo wpa_cli -i $iface set_network $netid auth_alg OPEN";

**Reply by billz** [reply]
Please be patient. A fix is in progress https://github.com/RaspAP/raspap-webgui/issues/2008

**Reply by billz** [reply]
Resolved in the latest release, along with several other WiFi client enhancements.
